ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+poossible we will need to describe how to prepare the arguments for a
+such a call from AVM code, rather than as transaction encoding.
+
+
+## Encoding
+
+This section describes how ABI types can be represented as byte strings.
+
+Like the [EthereumABI](https://docs.soliditylang.org/en/v0.8.6/abi-spec.html), this
+encoding specification is designed to have the following two
+properties:
+
+
+1. The number of non-sequential "reads" necessary to access a value is
+   at most the depth of that value inside the encoded array
+   structure. For example, at most 4 reads are needed to retrieve a
+   value at `a[i][k][l][r]`.
+2. The encoding of a value or array element is not interleaved with
+   other data and it is relocatable, i.e. only relative “addresses”
+   (indexes to other parts of the encoding) are used.
+
+
+### Types
+
+The following types are supported in the Algorand ABI.
+
+* `uint<N>`: An `N`-bit unsigned integer, where `8 <= N <= 512` and `N % 8 = 0`.
+* `byte`: An alias for `uint8`.
+* `bool`: A boolean value that is restricted to either 0 or 1. When encoded, up to 8 consecutive `bool` values will be packed into a single byte.
+* `ufixed<N>x<M>`: An `N`-bit unsigned fixed-point decimal number with precision `M`, where `8 <= N <= 512`, `N % 8 = 0`, and `0 < M <= 160`, which denotes a value `v` as `v / (10^M)`.
+* `<type>[<N>]`: A fixed-length array of length `N`, where `N >= 0`. `type` can be any other type.
+* `address`: Used to represent a 32-byte Algorand address. This is equivalent to `byte[32]`.
+* `<type>[]`: A variable-length array. `type` can be any other type.
+* `string`: A variable-length byte array (`byte[]`) assumed to contain UTF-8 encoded content.
+* `(T1,T2,...,TN)`: A tuple of the types `T1`, `T2`, …, `TN`, `N >= 0`.
+
+
+### Static vs Dynamic Types
+
+For encoding purposes, the types are divided into two categories: static and dynamic.
+
+The dynamic types are:
+
+*  `<type>[]` for any `type`
+    * This includes `string` since it is an alias for `byte[]`.
+* `<type>[<N>]` for any dynamic `type`
+* `(T1,T2,...,TN)` if `Ti` is dynamic for some `1 <= i <= N`
+
+All other types are static. For a static type, all encoded values of
+that type have the same length, irrepective of their actual value.
+
+
+### Encoding Rules
+
+Let `len(a)` be the number of bytes in the binary string `a`. The
+returned value shall be considered to have the ABI type `uint16`.
+
+Let `enc` be a mapping from values of the ABI types to binary
+strings. This mapping defines the encoding of the ABI.
+
+For any ABI value `x`, we recursively define `enc(x)` to be as follows:
+
+* If `x` is a tuple of `N` types, `(T1,T2,...,TN)`, where `x[i]` is the value at index `i`, starting at 1:
+    * `enc(x) = head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[N])`
+    * Let `head` and `tail` be mappings from values in this tuple to binary strings. For each `i` such that `1 <= i <= N`, these mappings are defined as:
```

Comment:
This section is based heavily on Ethereum's ABI, which also uses the `head`/`tail` notation, but your point is taken that this section could use some additional clarification to stand on its own better.

In response to the proposal for storing the entire static segment in arg 2 and dynamic segment in arg 3, keep in mind that the relationship between the statically- and dynamically-sized segments is hierarchical, since the dynamic segment again may consist of another pair of statically- and dynamically-sized segments.

Here's an example to illustrate this:
1. Say I have the type `[]byte` with value `['a', 'b', 'c']`. This would get encoded as:
```
uint16(3) byte('a') byte('b') byte('c')
[ static ][          dynamic          ]
```

2. Say I have the type `(byte[], byte, byte)` with value `(['a', 'b', 'c'], 'd', 'e')`. This would get encoded as:
```
uint16(2) byte('d') byte('e') uint16(3) byte('a') byte('b') byte('c')
[            static         ][               dynamic                ] <-- first layer
                             [ static ][          dynamic           ] <-- second layer
```

Also, what do you mean by "size limits on arguments individually and totally"?

Reply:
I think that Ethereum ABI is very complex and not a model to follow. It's not like you're going to reuse the insanely complex ABI encoding code used by Web3 or ethers.

I agree that in your system it is hierarchical, but I think that is confusing and hard to generate and deal with. I think it would be better to just have the dynamically sized values be heap pointers. For example,
```
Example 1:
static: uint16(0) // pointer
heap: uint16(3) a b c // size, then data, so a "Pascal string"

Example 2:
static: uint16(0) d e // pointer then data
heap: uint16(3) a b c // same data as before
```

Dynamically sized things as pointers and static things as inline is very simple to explain.

Programs can precompute the location of every static thing and pointer and then just do `txna ApplicationArgs 3 ; txna ApplicationArgs 2 ; read16BitsAtOffset static ; extract call` that pattern would then be extremely common and could be optimized into a single tight operation with 1 byte and 1+n parameters (static and then the extract parameters) rather than 8 bytes and 1+n parameters.

By size limits, I mean that the MaxAppTotalArgLen is 2048 --- https://github.com/algorand/go-algorand/blob/master/config/consensus.go#L848 --- and strings are at most 4096 --- https://github.com/algorand/go-algorand/blob/master/data/transactions/logic/eval.go#L51 --- so you know a lot about how big these things are going to be and how they could be constrained. For example, that's why we know the sizes always fit in 16 bits --- they actually fit in 11 bits, so you could, for example say they are 16 bits and use the extra 5 bits for type tags and size (i.e. you could specify the extremely common case of a string's length being 8 bits and save a byte there.) The ABI should say something about these limits.

As an aside, I find it very annoying that there's this weird difference between MaxAppTotalArgLen, MaxStringSize, and MaxAppSumKeyValueLens * MaxGlobalSchemaEntries, because they are 2k, 4k, and 8k (although you lose 64 bytes to the key.) The last difference is especially annoying because it means that I can't have a string that represents the entire global memory. (Reach uses one byte keys---0 through 63---and treats all of the global keys as 127 byte pages of big data structure. It's really annoying that I can't just treat that as an actual 8k memory region.)